### PR TITLE
Issue 596: Add profiler path to configuration

### PR DESCRIPTION
**Pull request by hennna**
Step 1
  - Added profiler_path_ member variable to AdminImpl class. Allows profiler path to be configurable rather than hard coded.
  - Return Http::Code::BadRequest when call to Cpu::startProfiler(...) fails.

Step 2 
 - Update admin profiler_path JSON configuration 

Issue link: https://github.com/lyft/envoy/issues/596 

@htuch for comments.

### CODE REVIEW

File: include/envoy/server/configuration.h [hunk 1]
```diff
@@ -147,6 +147,11 @@ class Admin {
   virtual const std::string& accessLogPath() PURE;
 
   /**
+   * @return const std::string& the admin profiler path.
```
Comment (htuch) [inline include/envoy/server/configuration.h]: Actually, it might be more semantically correct to call this profilePath (and similar renames elsewhere). It's where the profile is emitted, not the profiler itself. The comment might then be "@return const std::string& profiler output path."

File: source/common/json/config_schemas.cc [hunk 2]
```diff
@@ -899,6 +899,7 @@ const std::string Json::Schema::TOP_LEVEL_CONFIG_SCHEMA(R"EOF(
         "type" : "object",
         "properties" : {
           "access_log_path" : {"type" : "string"},
+          "profiler_path" : {"type" : "string"},
```
Comment (htuch) [inline source/common/json/config_schemas.cc]: Ditto etc.

Comment (htuch) [inline source/common/json/config_schemas.cc]: When you change `config_schemas.cc`, it's usually also necessary to update the user configuration docs. In this case, the file to update is https://github.com/lyft/envoy/blob/master/docs/configuration/overview/admin.rst.

File: source/common/profiler/profiler.h [hunk 3]
```diff
@@ -15,7 +15,7 @@ class Cpu {
   /**
    * Start the profiler and write to the specified path.
    */
-  static void startProfiler(const std::string& output_path);
+  static bool startProfiler(const std::string& output_path);
```
Comment (htuch) [inline source/common/profiler/profiler.h]: Add some return value documentation.

File: source/common/json/config_schemas.cc (hunk 2 above)
Comment (htuch) [inline source/common/json/config_schemas.cc]: You can also defer the `config_schemas.cc` change to your next PR, since it's not used yet here.

File: source/server/http/admin.cc [hunk 4]
```diff
@@ -317,7 +321,7 @@ AdminImpl::AdminImpl(const std::string& access_log_path, Network::Address::Insta
           {"/server_info", "print server version/status information",
            MAKE_HANDLER(handlerServerInfo)},
           {"/stats", "print server stats", MAKE_HANDLER(handlerStats)}} {
-
+  profiler_path_ = profiler_path;
```
Comment (htuch) [inline source/server/http/admin.cc]: It might be nicer to initialize this in the above constructor member initialization list, i.e.

``` : server_(server), ..., profiler_path_(profiler_path)```

File: test/integration/integration_admin_test.cc [hunk 5]
```diff
@@ -105,10 +105,12 @@ TEST_F(IntegrationTest, Admin) {
   EXPECT_TRUE(response->complete());
   EXPECT_STREQ("400", response->headers().Status()->value().c_str());
 
+  // TO DO: Change to expect 200 when profiler path is a configurable parameter
```
Comment (htuch) [inline test/integration/integration_admin_test.cc]: The convention used in Envoy is `TODO(hennna): foo bar`.

File: source/server/http/admin.cc [hunk 6]
```diff
@@ -163,7 +163,11 @@ Http::Code AdminImpl::handlerCpuProfiler(const std::string& url, Buffer::Instanc
 
   bool enable = query_params.begin()->second == "y";
   if (enable && !Profiler::Cpu::profilerEnabled()) {
-    Profiler::Cpu::startProfiler("/var/log/envoy/envoy.prof");
+    if (!Profiler::Cpu::startProfiler(profiler_path_)) {
+      response.add("?enable=<y|n>\n");
+      return Http::Code::BadRequest;
```
Comment (htuch) [inline source/server/http/admin.cc]: This returns a 400. I think it's not actually a bad client formed request, it's an internal server error, so you want to return a 500.

File: source/server/http/admin.cc [hunk 7]
```diff
@@ -163,7 +163,11 @@ Http::Code AdminImpl::handlerCpuProfiler(const std::string& url, Buffer::Instanc
 
   bool enable = query_params.begin()->second == "y";
   if (enable && !Profiler::Cpu::profilerEnabled()) {
-    Profiler::Cpu::startProfiler("/var/log/envoy/envoy.prof");
+    if (!Profiler::Cpu::startProfiler(profiler_path_)) {
+      response.add("?enable=<y|n>\n");
```
Comment (htuch) [inline source/server/http/admin.cc]: I think a more meaningful response would indicate there was a failure to start the profiler.

File: test/server/http/admin_test.cc [hunk 8]
```diff
@@ -13,13 +14,17 @@ class AdminFilterTest : public testing::Test {
...
+               server_),
+        admin_bad_profiler_path_("/dev/null", "/var/log/envoy/envoy.prof",
+                                 Network::Utility::resolveUrl("tcp://127.0.0.1:9002"), server_),
         filter_(admin_), request_headers_{{":path", "/"}} {
     filter_.setDecoderFilterCallbacks(callbacks_);
   }
 
   NiceMock<MockInstance> server_;
   AdminImpl admin_;
+  AdminImpl admin_bad_profiler_path_;
```
Comment (htuch) [inline test/server/http/admin_test.cc]: I don't think this needs to be a member of `AdminFilterTest`, it can be just allocated and instantiated locally in `AdminBadProfiler`.

File: test/server/http/admin_test.cc [hunk 9]
```diff
@@ -13,13 +14,17 @@ class AdminFilterTest : public testing::Test {
 public:
   // TODO(mattklein123): Switch to mocks and do not bind to a real port.
   AdminFilterTest()
-      : admin_("/dev/null", Network::Utility::resolveUrl("tcp://127.0.0.1:9002"), server_),
+      : admin_("/dev/null", "/tmp/envoy.prof", Network::Utility::resolveUrl("tcp://127.0.0.1:9002"),
```
Comment (htuch) [inline test/server/http/admin_test.cc]: This isn't a fault in this PR (it's a widespread issue in Envoy today), but using `/tmp` is kind of dangerous, since you may conflict with other users or test runs in a shared environment. You can keep the code as is, since we don't have a better solution yet, but add a `TODO(htuch): Use proper temporary path allocation method.` here?

Comment (htuch) [inline test/server/http/admin_test.cc]: I've filed https://github.com/lyft/envoy/issues/605 to track this work.

**Comment by mattklein123** [comment]
@hennna please sign CLA if you didn't already.

**Comment by hennna** [comment]
@mattklein123 just signed CLA

File: source/server/http/admin.cc [hunk 10]
```diff
@@ -317,7 +322,6 @@ AdminImpl::AdminImpl(const std::string& access_log_path,
           {"/server_info", "print server version/status information",
            MAKE_HANDLER(handlerServerInfo)},
           {"/stats", "print server stats", MAKE_HANDLER(handlerStats)}} {
-
```
Comment (htuch) [inline source/server/http/admin.cc]: Nit: unrelated line deletion?

File: source/common/profiler/profiler.h [hunk 11]
```diff
@@ -14,8 +14,9 @@ class Cpu {
 
   /**
    * Start the profiler and write to the specified path.
+   * @return bool whether the call to start profiler succeeded.
```
Comment (mattklein123) [inline source/common/profiler/profiler.h]: nit: "start the profiler"

File: source/server/http/admin.h [hunk 12]
```diff
@@ -113,6 +113,7 @@ class AdminImpl : public Admin,
 
   Server::Instance& server_;
   std::list<Http::AccessLog::InstanceSharedPtr> access_logs_;
+  std::string profile_path_;
```
Comment (mattklein123) [inline source/server/http/admin.h]: const

File: test/integration/integration_admin_test.cc [hunk 13]
```diff
@@ -105,28 +105,39 @@ TEST_F(IntegrationTest, Admin) {
...
   EXPECT_STREQ("200", response->headers().Status()->value().c_str());
 
-  response = IntegrationUtil::makeSingleRequest(ADMIN_PORT, "GET", "/cpuprofiler?enable=n", "",
+  response = IntegrationUtil::makeSingleRequest(ADMIN_PORT, "GET", "/reset_counters", "",
                                                 Http::CodecClient::Type::HTTP1);
   EXPECT_TRUE(response->complete());
   EXPECT_STREQ("200", response->headers().Status()->value().c_str());
 
-  response = IntegrationUtil::makeSingleRequest(ADMIN_PORT, "GET", "/hot_restart_version", "",
+  response = IntegrationUtil::makeSingleRequest(ADMIN_PORT, "GET", "/certs", "",
```
Comment (htuch) [inline test/integration/integration_admin_test.cc]: What's going on with all these handler path changes? They don't seem related.

File: test/integration/integration_admin_test.cc [hunk 14]
```diff
@@ -105,28 +105,39 @@ TEST_F(IntegrationTest, Admin) {
...
   EXPECT_TRUE(response->complete());
   EXPECT_STREQ("200", response->headers().Status()->value().c_str());
+}
 
-  response = IntegrationUtil::makeSingleRequest(ADMIN_PORT, "GET", "/reset_counters", "",
+// Successful call to startProfiler requires tcmalloc.
+#ifdef TCMALLOC
+
+TEST_F(IntegrationTest, AdminCpuProfilerStart) {
+  BufferingStreamDecoderPtr response = IntegrationUtil::makeSingleRequest(
```
Comment (mattklein123) [inline test/integration/integration_admin_test.cc]: I don't think you meant to move this request ("/") inside here. Should go back in other section.

File: test/server/http/admin_test.cc [hunk 15]
```diff
@@ -45,4 +48,27 @@ TEST_F(AdminFilterTest, Trailers) {
...
+  EXPECT_TRUE(Profiler::Cpu::profilerEnabled());
+  admin_.runCallback("/cpuprofiler?enable=n", data);
+  EXPECT_FALSE(Profiler::Cpu::profilerEnabled());
+}
+
+#endif
+
+TEST_F(AdminFilterTest, AdminBadProfiler) {
+  Buffer::OwnedImpl data;
+  AdminImpl admin_bad_profile_path("/dev/null", "/var/log/envoy/envoy.prof",
```
Comment (mattklein123) [inline test/server/http/admin_test.cc]: This is not going to fail in all envs. I would use a totally bogus path here.

Comment (htuch) [inline test/server/http/admin_test.cc]: /var/log/envoy/envoy.prof might be a valid path in some environments. Maybe use `/some/unlikely/bad/path.prof` for now, we should deal with this better when cleaning up the temp file story. Another approach would be to mock file handling, but I think this is a bit of clutter and overhead to put this dependency injection into AdminImpl.

File: test/integration/integration_admin_test.cc (hunk 13 above)
Comment (hennna) [inline test/integration/integration_admin_test.cc]: I moved the two enable=y and enable=n into it's own integration test called AdminCpuProfilerStart. I think this has to do with line order changes.
